tools/keepalived/keepalived/libipvs-2.6/libipvs.c

```diff
@@ -611,55 +664,63 @@ struct dp_vs_blklst_conf_array *ipvs_get_blklsts(void)
+	struct ip_vs_dest_entry *ipvs_entry;
+	struct dp_vs_dest_entry *dpvs_entry;
 	size_t len, len_rcv;
 	int i;
 
-	len = sizeof(*d) + sizeof(ipvs_dest_entry_t) * svc->num_dests;
-	if (!(d = malloc(len)))
+	len = sizeof(struct ip_vs_get_dests) + 
+		sizeof(ipvs_dest_entry_t) * svc->num_dests;
+	if (!(d = calloc(len, 1)))
 		return NULL;
 
-	ipvs_func = ipvs_get_dests;
-
-	if (svc->af != AF_INET) {
-	  errno = EAFNOSUPPORT;
-	  free(d);
-	  return NULL;
-	}
-
-	len = sizeof(*dk) + sizeof(struct ip_vs_dest_entry_kern) * svc->num_dests;
-	if (!(dk = malloc(len))) {
+	len = sizeof(struct dp_vs_get_dests) +
+		sizeof(struct dp_vs_dest_entry) * svc->num_dests;
+	if (!(dpvs_dests = calloc(len, 1))) {
 		free(d);
 		return NULL;
 	}
 
-	dk->fwmark = svc->fwmark;
-	dk->protocol = svc->protocol;
-	dk->addr = svc->addr.ip;
-	dk->port = svc->port;
-	dk->num_dests = svc->num_dests;
-	snprintf(dk->srange, sizeof(dk->srange), "%s", svc->srange);
-	snprintf(dk->drange, sizeof(dk->drange), "%s", svc->drange);
-	snprintf(dk->iifname, sizeof(dk->iifname), "%s", svc->iifname);
-	snprintf(dk->oifname, sizeof(dk->oifname), "%s", svc->oifname);
-
-	if (dpvs_getsockopt(DPVS_SO_GET_DESTS, dk, len, (void **)&dk_rcv, &len_rcv) < 0) {
+	dpvs_dests->af = svc->af;
+	dpvs_dests->fwmark = svc->fwmark;
+	dpvs_dests->proto = svc->protocol;
+	memcpy(&dpvs_dests->addr, &svc->addr, sizeof(svc->addr));
+	dpvs_dests->port = svc->port;
+	dpvs_dests->num_dests = svc->num_dests;
+	snprintf(dpvs_dests->srange, sizeof(dpvs_dests->srange), "%s", svc->srange);
+	snprintf(dpvs_dests->drange, sizeof(dpvs_dests->drange), "%s", svc->drange);
+	snprintf(dpvs_dests->iifname, sizeof(dpvs_dests->iifname), "%s", svc->iifname);
+	snprintf(dpvs_dests->oifname, sizeof(dpvs_dests->oifname), "%s", svc->oifname);
+
+	if (dpvs_getsockopt(DPVS_SO_GET_DESTS, dpvs_dests, len, (void **)&dpvs_dests_rcv, &len_rcv) < 0) {
 		free(d);
-		free(dk);
-        dpvs_sockopt_msg_free(dk_rcv);
+		free(dpvs_dests);
 		return NULL;
 	}
-	memcpy(d, dk_rcv, sizeof(struct ip_vs_get_dests_kern));
-	d->af = AF_INET;
-	d->addr.ip = d->__addr_v4;
-	for (i = 0; i < dk_rcv->num_dests; i++) {
-		memcpy(&d->entrytable[i], &dk_rcv->entrytable[i],
-		       sizeof(struct ip_vs_dest_entry_kern));
-		d->entrytable[i].af = AF_INET;
-		d->entrytable[i].addr.ip = d->entrytable[i].__addr_v4;
+
+	d->af = dpvs_dests_rcv->af;
+	memcpy(&d->addr, &dpvs_dests_rcv->addr, sizeof(d->addr));
+	d->protocol  = dpvs_dests_rcv->proto;
+	d->port = dpvs_dests_rcv->port;
+	d->fwmark = dpvs_dests_rcv->fwmark;
+	d->num_dests = dpvs_dests_rcv->num_dests;
+	snprintf(d->srange, sizeof(dpvs_dests_rcv->srange), "%s", dpvs_dests_rcv->srange);
+	snprintf(d->drange, sizeof(dpvs_dests_rcv->drange), "%s", dpvs_dests_rcv->drange);
+	snprintf(d->iifname, sizeof(dpvs_dests_rcv->iifname), "%s", dpvs_dests_rcv->iifname);
+	snprintf(d->oifname, sizeof(dpvs_dests_rcv->oifname), "%s", dpvs_dests_rcv->oifname);
+	if (d->af == AF_INET) {
+		d->__addr_v4 = d->addr.ip;
```

Comment:
IPv6 not support here?

Reply:
ipv4 should set __addr_v4; ipv4 show always read them instead of addr;